thunder/executors/nvmathex.py

```diff
@@ -0,0 +1,194 @@
+import logging
+import weakref
+from dataclasses import dataclass
+from functools import lru_cache, partial
+from collections.abc import Callable
+
+import nvmath
+
+import torch
+from nvmath.linalg.advanced import Matmul, MatmulOptions
+from torch.utils.benchmark import Timer
+
+from thunder.core import prims
+from thunder.core.proxies import TensorProxy
+from thunder.dynamo.benchmark_utils import TorchProfileTimer
+from thunder.extend import OperatorExecutor, register_executor
+
+logger = logging.getLogger(__name__)
+
+nvmath_matmul_ex: OperatorExecutor = OperatorExecutor("nvmath_matmul")
+register_executor(nvmath_matmul_ex)
+
+
+@dataclass(frozen=True, slots=True)
+class TensorDescriptor:
+    """
+    A dataclass to store the shape, stride, dtype, and device index of a tensor for caching purposes.
+    """
+
+    shape: tuple[int, ...]
+    stride: tuple[int, ...]
+    dtype: torch.dtype
+    device_index: int
+
+
+def execute_nvmath_matmul(mm: Matmul, a: torch.Tensor, b: torch.Tensor) -> torch.Tensor:
+    """
+    Executes a matrix multiplication operation using nvmath for the given operands and matmul executor.
+
+    Args:
+        mm (Matmul): The matmul executor.
+        a (torch.Tensor): The first operand.
+        b (torch.Tensor): The second operand.
+
+    Returns:
+        torch.Tensor: The result of the matrix multiplication operation.
+    """
+    mm.reset_operands(a, b)  # This function has about 10 µs overhead
+    return mm.execute()  # This function has about 75 µs overhead
+
+
+def nvmath_or_pytorch_matmul(mm: Matmul) -> Callable:
```

Comment:
fyi @kiya00

We should look at generalizing selecting executors by benchmarking